sympy/solvers/solveset.py

```diff
@@ -3665,33 +3665,34 @@ def nonlinsolve(system, *symbols):
         return _solveset_work(system, symbols)
 
     # main code of def nonlinsolve() starts from here
+    nonconstant = [eqn for eqn in map(_sympify, system) if eqn.free_symbols & set(symbols)]
     polys, polys_expr, nonpolys, denominators = _separate_poly_nonpoly(
-        system, symbols)
-
-    if len(symbols) == len(polys):
-        # If all the equations in the system are poly
+        nonconstant, symbols)
+    # first solve all polynomial equations:
+    if polys:
         if is_zero_dimensional(polys, symbols):
             # finite number of soln (Zero dimensional system)
             try:
-                return _handle_zero_dimensional(polys, symbols, system)
+                poly_res =  _handle_zero_dimensional(polys, symbols, nonconstant)
             except NotImplementedError:
                 # Right now it doesn't fail for any polynomial system of
                 # equation. If `solve_poly_system` fails then `substitution`
                 # method will handle it.
-                result = substitution(
+                poly_res = substitution(
                     polys_expr, symbols, exclude=denominators)
-                return result
-
-        # positive dimensional system
-        res = _handle_positive_dimensional(polys, symbols, denominators)
-        if res is S.EmptySet and any(not p.domain.is_Exact for p in polys):
-            raise NotImplementedError("Equation not in exact domain. Try converting to rational")
         else:
-            return res
-
+            # positive dimensional system
+            poly_res = _handle_positive_dimensional(polys, symbols, denominators)
+        if poly_res is S.EmptySet:
+            if any(not p.domain.is_Exact for p in polys):
+                return substitution(polys_expr + nonpolys, symbols, exclude=denominators)
+            else:
+                return S.EmptySet
+        if not nonpolys:
+            return poly_res
+        poly_sol = [dict(zip(symbols, sol)) for sol in poly_res.args]
+        return substitution(nonpolys, symbols, result=poly_sol, exclude=denominators)
+    # If all the equations are not polynomial.
+    # Use `substitution` method for the system
     else:
-        # If all the equations are not polynomial.
-        # Use `substitution` method for the system
-        result = substitution(
-            polys_expr + nonpolys, symbols, exclude=denominators)
-        return result
+        return substitution(nonpolys, symbols, exclude=denominators)
```

Comment:
Is this branch still needed?

Maybe just always solve the polynomial equations and pass the result to `substitution`.

Reply:
I think this is still needed because if there are no polynomial equations, that part of `nonlinsolve` which solves polynomial equations returns `EmptySet`, causing `nonlinsolve` to also return `EmptySet` even before `substitution` is called. So, the case where `polys` is empty has to be handled separately.

Reply:
This case could be absorbed into another condition though. I'll make the change.

Reply:
Now that `nonlinsolve` converts polynomial coefficients to `Rational` first, I don't think this branch can go anywhere else, so I think it's still needed as-is.